server/src/main/java/com/linecorp/centraldogma/server/internal/api/GitHttpService.java

```diff
@@ -0,0 +1,204 @@
+        if (repoName.length() > 5 && repoName.endsWith(".git")) {
+            repoName = repoName.substring(0, repoName.length() - 4);
+        }
+        return repoName;
+    }
+
+    // https://www.git-scm.com/docs/gitprotocol-http#_smart_service_git_upload_pack
+    @Post("/{projectName}/{repoName}/git-upload-pack")
+    public HttpResponse gitUploadPack(AggregatedHttpRequest req,
+                                      @Param String projectName, @Param String repoName) {
+        repoName = maybeRemoveGitSuffix(repoName);
+        final String gitProtocol = req.headers().get("git-protocol");
+        if (gitProtocol == null || !gitProtocol.contains(VERSION_2_REQUEST)) {
+            return HttpResponse.of(HttpStatus.BAD_REQUEST, MediaType.PLAIN_TEXT_UTF_8,
+                                   "Unsupported git-protocol: " + gitProtocol);
+        }
+
+        final MediaType contentType = req.headers().contentType();
+        if (contentType == null || !"application/x-git-upload-pack-request".equals(contentType.toString())) {
+            return HttpResponse.of(HttpStatus.BAD_REQUEST, MediaType.PLAIN_TEXT_UTF_8,
+                                   "Unsupported content-type: " + contentType);
+        }
+
+        if (!projectApiManager.exists(projectName)) {
+            return HttpResponse.of(HttpStatus.NOT_FOUND, MediaType.PLAIN_TEXT_UTF_8,
+                                   "Project not found: " + projectName);
+        }
+        if (!projectApiManager.getProject(projectName).repos().exists(repoName)) {
+            return HttpResponse.of(HttpStatus.NOT_FOUND, MediaType.PLAIN_TEXT_UTF_8,
+                                   "Repository not found: " + repoName);
+        }
+
+        final Repository jGitRepository =
+                projectApiManager.getProject(projectName).repos().get(repoName).jGitRepository();
+
+        final ByteStreamMessage body = StreamMessage.fromOutputStream(os -> {
+            // Don't need to close the input stream.
+            final ByteArrayInputStream inputStream = new ByteArrayInputStream(req.content().byteBuf().array());
+            // Don't need to close because we don't use the timer inside it.
+            final UploadPack uploadPack = new UploadPack(jGitRepository);
+            uploadPack.setTimeout(0); // Disable timeout because Armeria server will handle it.
+            // HTTP does not use bidirectional pipe.
+            uploadPack.setBiDirectionalPipe(false);
+            uploadPack.setExtraParameters(ImmutableList.of(VERSION_2_REQUEST));
+            try {
+                uploadPack.upload(inputStream, os, null);
+            } catch (IOException e) {
+                // Log until https://github.com/line/centraldogma/pull/719 is implemented.
+                logger.debug("Failed to respond git-upload-pack-request: {}", req, e);
+                throw new RuntimeException("failed to respond git-upload-pack-request: " + req, e);
```

Comment:
We seem to be using `verboseResponses`. I guess this could include the authorization header in the response which I'm not sure is a good idea.

Reply:
Modified to use just body. Thanks!